File: net/JetBrains.Serialization/src/Dataholders/MsCounterSignatureInfo.cs
using Newtonsoft.Json;
using Org.BouncyCastle.Asn1;
using JetBrains.Serialization;
using Org.BouncyCastle.Asn1.X509;

[JsonObject(MemberSerialization.OptIn)]
public class MsCounterSignatureInfo : IEncodableInfo
{
  [JsonProperty("Version")] private int _version;
  [JsonProperty("Algorithms")] private List<AlgorithmInfo> _algorithms;
  [JsonProperty("TstInfo")] private TSTInfo _tstInfo;

  [JsonProperty("TaggedCertificateInfo")]
  private TaggedObjectInfo _taggedCertificateInfo;

  [JsonProperty("CounterSignatures")] private List<CounterSignatureInfo> _counterSignatures;

  [JsonConstructor]
  public MsCounterSignatureInfo(
    int version,
    List<AlgorithmInfo> algorithms,
    TSTInfo tstInfo,
    TaggedObjectInfo taggedCertificateInfo,
    List<CounterSignatureInfo> counterSignatures)
  {
    _version = version;
    _algorithms = algorithms;
    _tstInfo = tstInfo;
    _taggedCertificateInfo = taggedCertificateInfo;
    _counterSignatures = counterSignatures;
  }

  public static MsCounterSignatureInfo GetInstance(DerSequence sequence)
  {
    var enumerator = sequence.GetEnumerator();

    enumerator.MoveNext();
    var version = ((DerInteger)enumerator.Current).Value.IntValue;

    enumerator.MoveNext();
    var algorithms = ((DerSet)enumerator.Current)
      .OfType<DerSequence>()
      .ToList()
      .Select(derSequence => new AlgorithmInfo(AlgorithmIdentifier.GetInstance(derSequence)))
      .ToList();

    enumerator.MoveNext();
    var tstInfo = new TSTInfo((DerSequence)enumerator.Current);

    enumerator.MoveNext();

    TaggedObjectInfo certificateInfoTagged = new TaggedObjectInfo(
      ((DerTaggedObject)enumerator.Current).IsExplicit(),
      ((DerTaggedObject)enumerator.Current).TagNo,
      new SequenceInfo(
        ((DerSequence)((DerTaggedObject)enumerator.Current).GetObject()).ToArray()
        .Select(it => CertificateInfo.GetInstance(it.ToAsn1Object())).ToList()
      )
    );

    enumerator.MoveNext();
    var counterSignatures = ((DerSet)enumerator.Current)
      .OfType<DerSequence>()
      .Select(CounterSignatureInfo.GetInstance)
      .ToList();

    return new MsCounterSignatureInfo(
      version,
      algorithms,
      tstInfo,
      certificateInfoTagged,
      counterSignatures);
  }

  public virtual Asn1Encodable ToPrimitive() =>
    new List<Asn1Encodable>
    {
      new DerInteger(_version),
      _algorithms.ToPrimitiveDerSet(),
      _tstInfo.ToPrimitive(),
      _taggedCertificateInfo.ToPrimitive(),
      _counterSignatures.ToPrimitiveDerSet()
    }.ToDerSequence();
}

File: jvm/src/main/kotlin/com/jetbrains/signatureverifier/serialization/dataholders/MsCounterSignatureInfo.kt
package com.jetbrains.signatureverifier.serialization.dataholders

import com.jetbrains.signatureverifier.serialization.toDLSequence
import com.jetbrains.signatureverifier.serialization.toPrimitiveDLSet
import kotlinx.serialization.Serializable
import org.bouncycastle.asn1.*
import org.bouncycastle.asn1.x509.AlgorithmIdentifier

@Serializable
data class MsCounterSignatureInfo(
  val version: Int,
  val algorithms: List<AlgorithmInfo>,
  val tstInfo: TSTInfo,
  val taggedCertificateInfo: TaggedObjectInfo,
  val counterSignatures: List<CounterSignatureInfo>

) : EncodableInfo {
  companion object {
    fun getInstance(sequence: DLSequence): MsCounterSignatureInfo {
      val iterator = sequence.iterator()

      val version = (iterator.next() as ASN1Integer).intValueExact()

      val algorithms = (iterator.next() as DLSet).map {
        AlgorithmInfo(
          AlgorithmIdentifier.getInstance(it)
        )
      }

      val tstInfo = TSTInfo(iterator.next() as DLSequence)

      val taggedCertificateInfo = iterator.next().let {
        TaggedObjectInfo(
          (it as DLTaggedObject).isExplicit,
          it.tagNo,
          SequenceInfo(
            (it.baseObject as DLSequence).map { certificateSequence ->
              CertificateInfo.getInstance(certificateSequence.toASN1Primitive())
            }
          )
        )
      }

      val counterSignatures = (iterator.next() as DLSet).map {
        CounterSignatureInfo.getInstance(it as DLSequence)
      }

      return MsCounterSignatureInfo(
        version,
        algorithms,
        tstInfo,
        taggedCertificateInfo,
        counterSignatures
      )
    }

  }

  override fun toPrimitive(): ASN1Primitive = listOf(
    ASN1Integer(version.toLong()),
    algorithms.toPrimitiveDLSet(),
    tstInfo.toPrimitive(),
    taggedCertificateInfo.toPrimitive(),
    counterSignatures.toPrimitiveDLSet()
  ).toDLSequence()
}